akvo/rsr/usecases/jobs/aggregation.py

```diff
@@ -0,0 +1,154 @@
+from __future__ import annotations
+
+import logging
+from typing import List, TYPE_CHECKING
+
+from django.db.models import QuerySet
+from django.db.transaction import atomic
+
+from akvo.utils import rsr_send_mail_to_users
+
+if TYPE_CHECKING:
+    from akvo.rsr.models import IndicatorPeriod
+
+from akvo.rsr.models.aggregation_job import IndicatorPeriodAggregationJob
+from akvo.rsr.models.cron_job import CronJobMixin
+from akvo.rsr.usecases.period_update_aggregation import aggregate
+from akvo.rsr.usecases.jobs.cron import is_job_dead
+
+logger = logging.getLogger(__name__)
+
+
+def get_scheduled_jobs() -> QuerySet[IndicatorPeriodAggregationJob]:
+    return base_get_jobs().filter(status=CronJobMixin.Status.SCHEDULED)
+
+
+def get_running_jobs() -> QuerySet[IndicatorPeriodAggregationJob]:
+    return base_get_jobs().filter(status=CronJobMixin.Status.RUNNING)
+
+
+def get_failed_jobs() -> QuerySet[IndicatorPeriodAggregationJob]:
+    return base_get_jobs().filter(status=CronJobMixin.Status.FAILED)
+
+
+def get_maxxed_jobs() -> QuerySet[IndicatorPeriodAggregationJob]:
+    return base_get_jobs().filter(status=CronJobMixin.Status.MAXXED)
+
+
+def get_finished_jobs() -> QuerySet[IndicatorPeriodAggregationJob]:
+    return base_get_jobs().filter(status=CronJobMixin.Status.FINISHED)
+
+
+def base_get_jobs() -> QuerySet[IndicatorPeriodAggregationJob]:
+    return IndicatorPeriodAggregationJob.objects.select_related("period__indicator")
+
+
+def schedule_aggregation_job(period: IndicatorPeriod) -> IndicatorPeriodAggregationJob:
+    """
+    Schedule a job for the period to be aggregated upwards if no job exists
+    """
+    logger.info("Scheduling indicator aggregation job for %s: %s", period, period.indicator.title)
+    if existing_job := get_scheduled_jobs().filter(period=period).first():
+        existing_job.save()
+        return existing_job
+
+    program = period.indicator.result.project.ancestor()
+    return IndicatorPeriodAggregationJob.objects.create(period=period, program=program)
+
+
+def execute_aggregation_jobs():
+    """
+    Call the aggregation function for each aggregation job
+    """
+    handle_failed_jobs()
+
+    scheduled_jobs = get_scheduled_jobs()
+    for scheduled_job in scheduled_jobs:
+        with atomic():
```

Comment:
I think this should be a while loop that keeps getting the first job. 

```py
while (scheduled_job := get_scheduled_jobs().first()):
```

And maybe `atomic` is unnecessary here too. Marking the job as running won't matter if the transaction ends because it will either be marked as finished or failed at the end of the transaction.